accepted/2021/package-validation/package-validation.md

```diff
@@ -0,0 +1,236 @@
+the previous version of the package:
+
+```XML
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net5.0</TargetFramework>
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <ValidatePackageAgainstPreviousVersion>True</ValidatePackageAgainstPreviousVersion>
+  </PropertyGroup>
+
+</Project>
+```
+
+A few weeks later, Skylar is tasked with adding support for a connection timeout
+to their library. The `Connect` method looks like this right now:
+
+```C#
+public static Client Connect(string url)
+{
+    // ...
+}
+```
+
+Since a connection timeout is an advanced configuration setting they reckon they
+can just add an optional parameter:
+
+```C#
+public static Client Connect(string url, TimeSpan? timeout = default)
+{
+    // ...
+}
+```
+
+However, when they rebuild the package they are getting an error:
+
+> error SYSLIB1234: The method 'Connect(string)' exists in the previous version
+> of the NuGet package (AdventureWorks.Client 1.2.3) but no longer exists in the
+> current version being built (1.2.4). This is a breaking change.
+
+Skylar realizes that while this is not a source breaking change, it's a binary
+breaking change. They solve this problem by adding an overload instead:
+
+```C#
+public static Client Connect(string url)
+{
+    return Client(url, Timeout.InfiniteTimeSpan);
+}
+
+public static Client Connect(string url, TimeSpan timeout)
+{
+    // ...
+}
+```
+
+### Making an intentional breaking change
+
+Armani is working on version 0.7 of their library. While they don't want to make
+accidental breaking changes, they do want to make changes based on user feedback.
+
+In version 0.6 they had a `Connect()` method that accepted a timeout as an
+`int`. Users complaint that timeouts in .NET are nowadays generally modelled
+using `TimeSpan` instead. After making the necessary code change, rebuilding the
+library results in an error:
+
+> error SYSLIB1234: The method 'Connect(string, int)' exists in the previous
+> version of the NuGet package (Fabrikam.Client 0.6.3) but no longer exists in
+> the current version being built (0.7.0). This is a breaking change.
+
+Since Armani's library is still an 0.x, which, according to SemVer, doesn't
+promise a stable API, they briefly consider turning off the breaking change
+validation entirely. However, Armani feels that even though it's 0.x they still
+don't want to make accidental breaking change so they decide to just suppress
+this specific instance of the breaking change:
+
+```C#
+[SuppressMessage("Interoperability", "SYSLIB1234")]]
+public static Client Connect(string url, TimeSpan timeout)
+{
+    // ...
+}
+```
+
+## Requirements
+
+### Goals
+
+* Performance should be good so that basic package validation can be on by
+  default (when the project generates a package), but it must be possible to
+  turn it off as well
+* Comparing to previous version should be opt-in because it requires downloading
```

Comment:
How would the tool work without comparing to a previous version?

Reply:
It would be running as part of the `dotnet pack` step and would make sure that the package is API Compatible (in between compile and runtime assets), well formed, closure complete, etc.

Reply:
@safern Will there be any flag or metadata that one can assess the `<ValidatePackageAgainstPreviousVersion>` property was used at pack time in the resulting package?